sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -2508,26 +2512,36 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
   pi_result Result;
   if (DeviceIsIntegrated) {
     Result = piextUSMHostAlloc(&Ptr, Context, nullptr, Size, Alignment);
+  } else if (Context->hasSingleRootDevice()) {
+    // If we have a single discrete device or all devices in the context are
+    // sub-devices of the same device then we can allocate on device
+    Result = piextUSMDeviceAlloc(&Ptr, Context, Context->SingleRootDevice,
+                                 nullptr, Size, Alignment);
   } else {
-    Result = piextUSMDeviceAlloc(&Ptr, Context, Context->Devices[0], nullptr,
-                                 Size, Alignment);
+    // Context with several gpu cards. Temporarily use host allocation because
+    // it is accessible by all devices. But it is not good in terms of
+    // performance.
+    // TODO: We need to either allow remote access to device memory using IPC,
+    // or do explicit memory transfers from one device to another using host
+    // resources as backing buffers to allow those transfers.
+    Result = piextUSMHostAlloc(&Ptr, Context, nullptr, Size, Alignment);
   }
+
   if (Result != PI_SUCCESS)
     return Result;
 
   if (HostPtr) {
     if ((Flags & PI_MEM_FLAGS_HOST_PTR_USE) != 0 ||
         (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) != 0) {
       // Initialize the buffer with user data
-      if (DeviceIsIntegrated) {
-        // Do a host to host copy
-        memcpy(Ptr, HostPtr, Size);
-      } else {
-
+      if (Context->hasSingleRootDevice() && !DeviceIsIntegrated) {
```

Comment:
I think the original check for `if (DeviceIsIntegrated) { memcpy } else { zeCommandListAppendMemoryCopy }` should stay here. As you have it now you would memcpy for multiple root-device contexts.